### PR TITLE
Do not apply environment maps to unlit models. Fixes #374

### LINKED ISSUE #374: Bad render results for KHR_materials_unlit Khronos test model
The [Khronos Unlit Test sample model](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/UnlitTest) README says that for the test model using this extension:

> The visible surface ... should appear as a uniform color, so much so that the 3D effect is intentionally lost.

And offers the following as examples of good and bad render results:

Good | Bad
------|-----
image|image

Here are examples of how `<model-viewer>` renders this model w/ various settings, all if which seem incorrect (possibly for different reasons):

model only|HDR `background-image`|`experimental-pmrem`
-----------|--------------------------|-----------------------
image|image|image

I also tested to see how Filament's `gltf_viewer` handles this case. It seems to handle the model correctly (presumably it supports the `KHT_materials_unlit` extension).

Filament `gltf_viewer`|
----------------------|
image|

### PR DESCRIPTION
**Pull request by jsantell**
Currently with a default env map, image env map, and PMREM'd env map currently:

localhost_5000_examples_fullscreen2 html 11

With this patch:

localhost_5000_examples_fullscreen2 html 10

Note that `UnlitTest.glb` is included (4kb) -- can reference the Khronos models if/when they land. The Astronaut is *not* unlit AFAICT (or, does not use the unlit extension), maybe should change the background-image demo describing it as such.

### CODE REVIEW

File: src/three-components/Model.js [hunk 1]
```diff
@@ -50,7 +50,10 @@ export default class Model extends Object3D {
 
   applyEnvironmentMap(map) {
     this.modelContainer.traverse(obj => {
-      if (obj && obj.isMesh && obj.material) {
+      // Unlit models (using MeshBasicMaterial) should not apply
+      // an environment map, even though `map` is the currently configured
+      // environment map.
+      if (obj && obj.isMesh && obj.material && !obj.material.isMeshBasicMaterial) {
```
Comment (donmccurdy) [inline src/three-components/Model.js]: For some edge cases, `obj.material` could be an array of materials. We should probably avoid creating those in GLTFLoader at some point, but for now it's worth handling.

**Comment by cdata** [comment]
@jsantell since we have merged #375 , would you be willing to update this change to use the unlit sample from the Khronos samples so that we don't have to check the model into our repo?

**Comment by jsantell** [comment]
@cdata  
> since we have merged #375 , would you be willing to update this change to use the unlit sample from the Khronos samples so that we don't have to check the model into our repo?

wouldn't that require needing to fetch all the models before being able to run tests? UnlitTest.glb is 4kb as well.

@donmccurdy 
> For some edge cases, `obj.material` could be an array of materials. We should probably avoid creating those in GLTFLoader at some point, but for now it's worth handling.

Any models that you know that do this to test?

**Comment by cdata** [comment]
>> since we have merged #375 , would you be willing to update this change to use the unlit sample from the Khronos samples so that we don't have to check the model into our repo?
> wouldn't that require needing to fetch all the models before being able to run tests? UnlitTest.glb is 4kb as well.

Yes, and we do this (models are fetched as part of `npm prepare`, so they happen after `npm install` and `npm ci`).

In practice this is probably fine. We have set up Travis to cache the directory they are checked out to, and it doesn't seem to take much time. We also update in place if we detect that the models are already fetched.

**Comment by cdata** [comment]
FYI looks like `UnlitTest.glb` is still checked in

**Comment by jsantell** [comment]
Using the Khronos sample models now, and handle when `material` is an array

File: src/three-components/Model.js [hunk 2]
```diff
@@ -49,10 +49,24 @@ export default class Model extends Object3D {
...
-        obj.material.envMap = map;
-        obj.material.needsUpdate = true;
+      if (obj && obj.isMesh && Array.isArray(obj.material)) {
+        materialStack = materialStack.concat(obj.material);
+      }
+      else if (obj && obj.isMesh && obj.material) {
+        materialStack.push(obj.material);
+      }
+
+      for (let material of materialStack) {
```
Comment (cdata) [inline src/three-components/Model.js]: Did you mean to do iteration of the material stack on every invocation of the `traverse` callback?

Comment (jsantell) [reply src/three-components/Model.js]: Mm, yes, but missing clearing it out. I don't like any way that attempts to not duplicate the logic for the (untested) edge case as well. Will change

**Comment by donmccurdy** [comment]
It's a bit contrived, but here's an example that will result in `mesh.material` being an array. The model contains a single mesh with two primitives, and the primitives are identical except for using different materials.

[Triangle.gltf.zip](https://github.com/GoogleWebComponents/model-viewer/files/2902385/Triangle.gltf.zip)

**Comment by jsantell** [comment]
* Fixed the material array logic
* Added Triangle.gltf for test case (added PBR values to see envmap), shook out some issues in our GLTFLoader wrapper
* Remove UnlitTest.glb

File: src/test/features/environment-spec.js [hunk 3]
```diff
@@ -160,6 +175,34 @@ suite('ModelViewerElementBase with EnvironmentMixin', () => {
...
+          });
+          element.src = UNLIT_MODEL_URL;
+          await onLoad;
+        });
+        test('applies no environment map on unlit model', async function() {
+          expect(modelHasEnvMap(scene)).to.be.false;
+        });
+      });
+
+      suite('on a model with multi-material meshes', () => {
```
Comment (cdata) [inline src/test/features/environment-spec.js]: 💯
